### ISSUE TITLE
Refine the Site Editor loading state

### ISSUE THREAD

**Issue by jameskoster** [WordPress/gutenberg#35503]
Here's the current experience loading the site editor:

https://user-images.githubusercontent.com/846565/136778002-4c8c7264-f37f-4b6f-abe8-524966350380.mp4

A couple of issues to address:

- [ ] The FOWPA (Flash of wp-admin) is awkward. Can we hide it?
- [x] The light grey background feels a bit broken. Can we use the dark grey background from the template editor, and isolated template part editor? https://github.com/WordPress/gutenberg/pull/35520
- [ ] The "Loading..." label in the top bar is easily missed. Something on the canvas would be more obvious.

**Comment by mtias** [comment]
Also the way different template parts can load at different times can be disorienting. I think we should consider loading everything before showing anything so that there are no jumps. Maybe also treat the gray as an overlay so we render in the background and then fade it out.

**Comment by talldan** [comment]
I created an issue for discussing the preloading of site editor content over here: https://github.com/WordPress/gutenberg/issues/35942

That would get rid of the disorienting nature of loading the editor. It might not be a quick win for 5.9, but could be something to consider in the longer term.

**Comment by jameskoster** [comment]
I worked on a few full-screen loading graphics [here](https://www.figma.com/proto/xndhKtrSkvL919bed0wwXc/Loader?page-id=91%3A1418&node-id=4%3A248&viewport=326%2C48%2C1&scaling=min-zoom&starting-point-node-id=4%3A248&show-proto-sidebar=1) and assembled my favorites in to some prototypes that demonstrate the full flow [here](https://www.figma.com/proto/xndhKtrSkvL919bed0wwXc/Loader?page-id=91%3A1419&node-id=27%3A728&viewport=326%2C48%2C1&scaling=scale-down&starting-point-node-id=27%3A728&show-proto-sidebar=1).

**Comment by shaunandrews** [comment]
I like the simplicity of the "Indeterminate bar" option, and I think it would work really well with Matías' suggestion of loading the pieces in bit-by-bit with an overlay.

There's something _off_ about the geometry of the "Layout" options; I think it has to do with the border radii and line-widths being different from the Gutenberg icon language. Perhaps it could work better if it felt more like a large icon, rather than its own thing.

--

I think there's an important bit to consider in this design: How does this handle things when it never actually loads for some reason?

**Comment by jameskoster** [comment]
I suppose it depends on what exactly doesn't load. Hypothetical scenarios:

* Nothing loading at all – total timeout
* Editor chrome (top bar, breadcrumbs, etc) loads – but nothing else
* Editor chrome + template loads, but template part(s) fail

I'm not sure how likely each of these are, or if I'm missing any?

Doesn't seem like there's much we could do in the first scenario. For the others, a modal / placeholder with "Try again" action might work?


https://user-images.githubusercontent.com/846565/140794104-a6f8ad0a-7160-4ae4-ac68-84d79789426e.mp4

**Comment by mtias** [comment]
@youknowriad and I discussed a broader issue of patterns and templates loading in stages that we should aim to address holistically. The end goal should be that things load like a "screenshot"; all fully loaded when you start interacting.

**Comment by shaunandrews** [comment]
> The end goal should be that things load like a "screenshot"; all fully loaded when you start interacting.

Not sure I know what this means... are you suggesting something like progressive loading, where things are loaded as a low-resolutions (literally and figuratively) at first, and detail is added over time until complete?

**Comment by mtias** [comment]
Sorry, I mean render the final state and don't show tearing (in progress, half-rendered states). Wait until everything is loaded.

**Comment by jameskoster** [comment]
If we're able to determine how many _things_ there are to load, and the status of each one, we could perhaps display a determinate loading indicator. Something like:


https://user-images.githubusercontent.com/846565/151210703-bc8ca029-2481-421c-a6ea-fa75ba3820f9.mp4



We may want to embellish it with a logo, or some accompanying text, but I kind of like the simplicity of the bar on its own.

Errors could be handled like so (with a better error message, obviously :)):


https://user-images.githubusercontent.com/846565/151210731-6b9aafdb-9827-4acc-a6e0-bbf1e83d53ef.mp4

**Comment by mtias** [comment]
@youknowriad @jsnajdr I believe the work on suspense at the data layer is in a place that can unlock working on this?

**Comment by youknowriad** [comment]
@mtias Yes, I think now that the API is in, the next step would be to add the SuspenseProvider to the editors (edit-post, edit-site...) and start switching some useSelects to this and see the impact.

**Comment by tyxla** [comment]
I tinkered a bit today with `useSuspenseSelect` with the site editor, here's a proof-of-concept:

#42525

I'm heading AFK until the end of next week, but when I'm back I'm happy to continue experimenting with this.

**Comment by annezazu** [comment]
@tyxla is this something you're still working on? Checking in considering this is one of the issues noted for removing the beta label :)

**Comment by tyxla** [comment]
Yes @annezazu  👍 I'm planning to do some further experimentation and refinement.

**Comment by getdave** [comment]
I'm curious about any next steps here. @tyxla is there anything specific that needs to be addressed or is it ongoing work spanning across WP releases?

**Comment by tyxla** [comment]
Thanks for asking @getdave - as mentioned [here](https://github.com/WordPress/gutenberg/pull/42525#issuecomment-1259286862) this experiment has been blocked by the React 18 upgrade that we're working on in https://github.com/WordPress/gutenberg/pull/45235. Once we've upgraded, we'd like to try with `startTransition` / `useTransition` and we're hoping it would solve the problems we're having, unblocking the experiment. If that fails, we could try an alternative, non-suspense solution that would essentially be looking at the resolving state of all the stores of interest - cc @jsnajdr as we've been discussing that recently.

**Comment by getdave** [comment]
Thanks for the update. I was curious from a Phase 2 perspective (i'm doing some research) as to whether this was still WIP. Looking forward to seeing the improvements 🤞

**Comment by mtias** [comment]
Yes, still very much an important part of phase 2.

**Comment by annezazu** [comment]
Noting that upgrading to React 18 was merged in https://github.com/WordPress/gutenberg/pull/45235, marking a big step forward to unblock this work!

**Comment by tyxla** [comment]
Latest update: In #46467 we're now working to enable concurrent rendering features, which is the last step before we're able to experiment with `startTransition` / `useTransition` and altering `useSelect` to use `useSyncExternalStore`.

**Comment by tyxla** [comment]
I've been working on another version of the experiment today: 

https://github.com/WordPress/gutenberg/pull/47612

Here I tried another approach, which I'm hoping to be better and more flexible in terms of us being able to declare the dependencies for each suspense boundary.

**Comment by talldan** [comment]
It seems like the post editor also needs some work now, possibly since the iframe was introduced:

https://user-images.githubusercontent.com/677833/216228500-6cb53e6f-96c1-4d3d-b186-e966d49b1d87.mp4

It might be a different problem though, it's more likely that the styles in the iframe are not ready immediately, and that's causing a flash of unstyled content.

I can make a separate issue if you think that's best.

**Comment by tyxla** [comment]
If you ask me, this is worth a separate issue. Not only that it's in the post content, but it might be solved with a different approach - for example by waiting for the iframe to fully load before showing it.

**Comment by talldan** [comment]
I've made an issue for it - https://github.com/WordPress/gutenberg/issues/47724.

**Comment by annezazu** [comment]
As WordPress 6.2 gets underway, I wanted to quickly offer a recap of the work that's been done on this issue and why it remains open for now. 

- A PR was merged to [Update the site editor for better initialization](https://github.com/WordPress/gutenberg/pull/47182#top).
- A PR was merged to [enable concurrent mode by implementing mount/unmount with createRoot](https://github.com/WordPress/gutenberg/pull/46467#top).

Finally, an additional PR is underway-[Site Editor: Improve loading experience](https://github.com/WordPress/gutenberg/pull/47612#top). For WordPress 6.2, meaningful progress in this area has been made where there's enough confidence to move forward with removing the beta label. Improving the loading state, similar to areas like performance, are iterative and continual in many ways. As a result, this issue is remaining open since there is more that _can_ be done with access to various APIs after the React 18 upgrade that can smooth the experience even more in an iterative manner for the next major WordPress release.

**Comment by tyxla** [comment]
Thanks for the recap, @annezazu 👍 More experimentation will come in this area in the following weeks, likely in the form of new PRs, as suggested [here](https://github.com/WordPress/gutenberg/pull/47612#issuecomment-1412231179) for example. Will keep everyone posted. We'll find a good way to improve the loading experience for everyone 👍

**Comment by annezazu** [comment]
👋🏼 Checking in here to see if [this experimental PR is still active](https://github.com/WordPress/gutenberg/pull/47612) and where work stands for the overall refining the site editor loading experience as it's been two months since the last update for 6.2 related items. I want to make sure I'm not missing anything in the phase 2 monthly updates :)

**Comment by tyxla** [comment]
Thanks for checking in @annezazu - it hasn't been a priority over the past weeks, but there's still some experimentation that we'd like to do, albeit it not being on top of our list.

**Comment by tyxla** [comment]
I've been working on a third version of the experiment, which I believe might be the one we end up using. 

Still need to polish it, fix tests, add demos, etc. but it looks promising so far: #50222.

**Comment by tyxla** [comment]
#50222 has just landed, improving the site editor loading a lot, particularly the editor canvas loading:

https://github.com/WordPress/gutenberg/assets/8436925/b7112e09-7b91-46ea-be78-938ebabc4314

We still have to work on improving the zoomed-out view loading state, so I suggest that we keep this issue open as we're working on that.

**Comment by mtias** [comment]
I'm fine closing this one since it's our main tracker for 6.3 and it's useful to see it's covered. We should break down followups and further improvements more granularly into a tracking issues.
